### DISCUSSION TITLE
Error in diversity(x, index = "shannon") : input data must be numeric

### DISCUSSION THREAD

**Discussion by Anibaba007** [vegandevs/vegan#481] [Q&A]
Hello everyone, 
I have tried running the analysis which aims to plot Shannon diversity index across land-use types. 
However, upon running the H line I got the error message:

**H <- apply(nativeAB, MARGIN = 1,FUN = diversity, index="shannon")
Error in FUN(newX[, i], ...) : input data must be numeric**

Below are my script and data attached. 

nativeAB = read.csv('all_transposedDR.csv')#Read data
head(nativeAB)#look at first 6 rows
str(nativeAB)

### A plot of Shannon-Wiener H' across the three land-use types
library(vegan)

H <- apply(nativeAB, MARGIN = 1,FUN = diversity, index="shannon")
H
CI<- apply(nativeAB, MARGIN = 1, FUN = H_CI)
CI
diversitybarplot <- barplot(shannon, ylim = c(0, max(H+CI)), cex.names = 0.8)
arrows(diversitybarplot, H+CI, diversitybarplot, H-CI, length = 0.1, angle = 90, code = 3)
title(xlab = "land use types", ylab = "Shannon-Wiener H' index")

[Copy of all_transposedDR.xlsx](https://github.com/vegandevs/vegan/files/5747136/Copy.of.all_transposedDR.xlsx)

**Comment by jarioksa** [comment]
I think the error message is clear: `input data must be numeric`. When I look at the Excel file that you sent, I see that input data are not numeric, and you should have seen the same when you issued `head(nativeAB)`. At least variables `site` and `land.use` are character strings (non-numeric) and also variables `transect`  and `plot`  should be analysed as non-numeric factors. You should omit those columns from your calculation. This is also very important: probably `transect` and `plot` will be read in as numeric and `diversity` can use them, but if you include them in your diversity calculations, your results are rubbish.

You can remove columns using negative indexing (and you do not need to use `apply`: `diversity` is automatically applied to each row):

```r
H <- diversity(nativeAB[, -(1:4)]) # removes cols 1 to 4, but check the actual indexing in your data
```

As a non-**vegan** issue: `site` and `land.use` are better treated as factors than character strings. However, for some obscure reason, **R** version 4 handles them as character strings. If you are using current **R**, you should add keyword `stringsAsFactors=TRUE`  in your `read.csv()` command. Then you should still change your numeric `transect` and `plot` to factors.

The last rows of your example are a mystery to me, as I have no idea what is function `H_Cl` or how you get something that you call `shannon` in your `barplot()` command.
